### PR TITLE
Implement execute_boundary_following() in robot.py

### PR DESCRIPTION
**Pull request by Aayush-Agnihotri**
### Description
Implemented dummy variables that will change (i.e. margin, turn_angle, and forward_dist) to provide baseline algorthm that can be tested and adjusted to find optimal values.

## Test Plan
In order to test this, we are creating a GUI to visualize the turns and explore edge cases as well as validate our main algorithm.

### CODE REVIEW

File: engine/robot.py [hunk 1]
```diff
@@ -44,9 +44,10 @@ class Robot:
     heading is in range [0..359]
     """
 
-    def __init__(self, x_pos, y_pos, heading, epsilon, max_v, radius, is_sim=True, position_kp=1, position_ki=0,
+    def __init__(self, x_pos, y_pos, heading, epsilon, max_v, radius, width, front_ultrasonic, lf_ultrasonic,
+                 lb_ultrasonic, rf_ultrasonic, rb_ultrasonic, is_sim=True, position_kp=1, position_ki=0,
```
Comment (AC408) [inline engine/robot.py:48]: Try setting a default value for these to see if the PR tests runs to completion (None for ultrasonic, 0 for width)

File: engine/robot.py [hunk 2]
```diff
@@ -330,10 +361,53 @@ def set_phase(self, new_phase):
...
-        # TODO: SET BACK TO ORIGINAL MISSION (TRAVERSE OR RETURN)
+    def track_obstacle(self):
         pass
 
+    def execute_avoid_obstacle(self, dist_to_goal, prev_phase):
+        pass
+
+    def execute_boundary_following(self, min_dist):
+        front_dist = self.rf_ultrasonic.distance()
+        back_dist = self.rb_ultrasonic.distance()
```
Comment (AC408) [inline engine/robot.py:372]: will there be support for boundary following involving getting the distance on the left side? Please comment which orientation the robot will traverse the obstacle with.

File: engine/robot.py [hunk 3]
```diff
@@ -330,10 +361,53 @@ def set_phase(self, new_phase):
...
+        margin = 1  # Will change
```
Comment (AC408) [inline engine/robot.py:373]: comments about what margin is or more descriptive naming will be great!

File: engine/robot.py [hunk 4]
```diff
@@ -330,10 +361,53 @@ def set_phase(self, new_phase):
...
+        forward_dist = 0.01  # Will change
+        turn_angle = math.pi / 90  # Will change
+
+        if math.abs(front_dist - back_dist) < margin:
```
Comment (AC408) [inline engine/robot.py:377]: I see that there's one value for turn_angle. Will there be support for turning the other way?

**Review by AC408 — CHANGES_REQUESTED** [review]
Good first steps! I think we can use this PR as an update kind of thing and merge with main once all the edge cases are flushed out.

**Review by AC408 — APPROVED** [review]

File: engine/robot.py [hunk 5]
```diff
@@ -330,10 +361,63 @@ def set_phase(self, new_phase):
...
+        pass
+
+    # Currently, algorithm only supports using the right side sensors of the robot for boundary following.
+    # When encountering an obstacle, robot will begin boundary following by going clockwise around the object.
+    # This was arbitrarily decided since currently there is no better way to determine which direction the robot should
+    # navigate around the obstacle.
+    def execute_boundary_following(self, min_dist):
+        front_dist = self.rf_ultrasonic.distance()
+        back_dist = self.rb_ultrasonic.distance()
+        margin = 1
```
Comment (sazhang02) [inline engine/robot.py:377]: how are we choosing this margin value?

File: engine/robot.py [hunk 6]
```diff
@@ -330,10 +361,63 @@ def set_phase(self, new_phase):
...
+            * Three cases:
+                1. Front right sensor is farther away from the obstacle than the back right sensor
+                    * Turn right until sensor inputs are equalized (front of the robot will turn closer toward the obstacle)
+                2. Front right sensor is closer to the obstacle than the back right sensor
+                    *  Turn left until sensor inputs are equalized (front of the robot will turn away from the obstacle)
+                3. Both right-side sensors displaying the same value
+                    * With a margin of error accounting for not uniform obstacles
+        2. Once both sensors are displaying the same values, move forward (boundary following)
+        3. Exits boundary following when conditions met in execute_avoid_obstacle() method
+        '''
```
Comment (sazhang02) [inline engine/robot.py:406]: Move this specification to directly under the function header? + merge with the comments above it?

**Review by sazhang02 — APPROVED** [review]
Nice!! Just added a question for my understanding and was thinking we can move the location of the comments. If you guys planned to merge, I think it would be fine after imo \o/

File: engine/robot.py (hunk 5 above)
Comment (Aayush-Agnihotri) [reply engine/robot.py:390]: Right now we set it as 1 arbitrarily, but in the future it will be the extra padding distance in meters for the width of the robot to ensure there are no problems boundary following (i.e. the side of the robot clipping the obstacle).

File: engine/robot.py [hunk 7]
```diff
@@ -48,7 +56,8 @@ class Robot:
     def __init__(self, x_pos, y_pos, heading, epsilon, max_v, radius, width=0, front_ultrasonic=None, lf_ultrasonic=None,
                  lb_ultrasonic=None, rf_ultrasonic=None, rb_ultrasonic=None, is_sim=True, position_kp=1, position_ki=0,
                  position_kd=0, position_noise=0, heading_kp=1, heading_ki=0, heading_kd=0, heading_noise=0,
-                 init_phase=1, time_step=1, move_dist=.5, turn_angle=3, plastic_weight=0, goal_location=(0, 0)):
```
Comment (sazhang02) [inline engine/robot.py]: did you mean to get rid of goal_location?

File: engine/robot.py [hunk 8]
```diff
@@ -405,7 +436,13 @@ def traverse_roomba(self, base_station_loc, time_limit, roomba_radius):
...
             dt += 1
             exit_boolean = (dt > time_limit)
-        self.phase = Phase.COMPLETE  # TODO: CHANGE the next phase to return
+            # if self.avoid_obstacle:
+            #     self.set_phase(Phase.AVOID_OBSTACLE)
+            #     self.prev_phase = Phase.TRAVERSE
+            #     return None
+        self.set_phase(Phase.COMPLETE)  # TODO: CHANGE the next phase to return
+        # self.phase = Phase.COMPLETE  # TODO: CHANGE the next phase to return
+
```
Comment (sazhang02) [inline engine/robot.py:445]:
```suggestion
```
delete extra set phase?

**Review by sazhang02 — CHANGES_REQUESTED** [review]
Two things!

**Review by sazhang02 — APPROVED** [review]
Nice!
